jlm/rvsdg/region.hpp

```diff
@@ -159,13 +160,128 @@ class Region
       intrusive_list<jlm::rvsdg::node, jlm::rvsdg::node::region_bottom_node_list_accessor>
           region_bottom_node_list;
 
+  using RegionArgumentIterator = std::vector<RegionArgument *>::iterator;
+  using RegionArgumentConstIterator = std::vector<RegionArgument *>::const_iterator;
+  using RegionArgumentRange = util::iterator_range<RegionArgumentIterator>;
+  using RegionArgumentConstRange = util::iterator_range<RegionArgumentConstIterator>;
+
+  using RegionResultIterator = std::vector<RegionResult *>::iterator;
+  using RegionResultConstIterator = std::vector<RegionResult *>::const_iterator;
+  using RegionResultRange = util::iterator_range<RegionResultIterator>;
+  using RegionResultConstRange = util::iterator_range<RegionResultConstIterator>;
+
+  using TopNodeIterator = region_top_node_list::iterator;
+  using TopNodeConstIterator = region_top_node_list::const_iterator;
+  using TopNodeRange = util::iterator_range<TopNodeIterator>;
+  using TopNodeConstRange = util::iterator_range<TopNodeConstIterator>;
+
+  using NodeIterator = region_nodes_list::iterator;
+  using NodeConstIterator = region_nodes_list::const_iterator;
+  using NodeRange = util::iterator_range<NodeIterator>;
+  using NodeConstRange = util::iterator_range<NodeConstIterator>;
+
+  using BottomNodeIterator = region_bottom_node_list::iterator;
+  using BottomNodeConstIterator = region_bottom_node_list::const_iterator;
+  using BottomNodeRange = util::iterator_range<BottomNodeIterator>;
+  using BottomNodeConstRange = util::iterator_range<BottomNodeConstIterator>;
```

Comment:
This is not completely clean as it exposes the underlying data structure to the outside world. It is for example visible when iterating through the arguments/results as I get pointers instead of references by default.

It is an improvement compared to before, but abstracting this away would have meant to implement our own iterator class in the region class, and I was too lazy for that. I will do it if you insist though.

Reply:
This is absolutely fine. There is a thing in C++ called `std::reference_wrapper`, which sits somewhere between pointers and references, which in theory could make it possible to store references in a vector, but I think the type is a bit scary, so I will gladly iterate over pointers.